### DISCUSSION TITLE
How to dynamically load and generate pages?

### DISCUSSION THREAD

**Discussion by kaosine** [vuepress/core#1112] [Q&A] [answered]
We have a database that people can browse using our api.pulsar-edit.dev/<mapping-point-to-import> and other common paths we want to be able to have users explore in a sane way:

> We would need pages for the following:
> * /packages - Multiple package view
> * /packages/:packageName - Single package view
> * /themes - Many package view that only shows themes
> * /themes/:packageName - Single package view of a specific theme
> * /users/:userName - User view. Not sure how useful it is, but this was 

We've tossed around using axio or superagent as a way to import it, but not sure how to provide dynamic routing for what it returns. Not only that, but we figured we could use a few things out of config, client config, and the node API but couldn't find the parts about the routing information to provide for that part to generate a page from that. Especially when it returns specific errors like 404, or any others. I know I would have to generate layouts for that, and that's easy enough. It's just figuring out the routing part from the data we get(in json).

**Comment by kaosine** [comment]
Any ideas on this? I found something through various pages, but none of it seemed to be what the person who is in charge of our API was looking for or seemed to explain it very well. We just need a way to dynamically generate and load pages from our api, so we can have people search for our various packages and themes for the editor.

**Comment by Mister-Hope** [comment, marked as answer]
If you want to render page "Pure dymanicly"

> What I mean is that you are trying to load api and register routes at runtime.

That means you are dropping vuepress SSG support. Any static HTML content will no longer be generated.

This means you have to deal 2 situation:

1. Handle redirect yourself like using other SPA. i.e.: you will get a 404 (not 404 page) when you visit a dynamic route directly, you should redirect them to `/404.html` or `/index.htm`
2. You should always detect current route path in layouts, if it fits some of your route patterns, do redirect to a loading page, then back to 404 if the api is not giving result, or register the route with layout you need and direct to the route if the api returns some content.


The problem is when you are acutally using Vue, you are using a SPA, and all resources should be redirected to an exisiting file. While vuepress do SSR at build time and provide SSG contents, so when you visit a package, you are loading that resources static content then do a client Hydration. So you no longer need redirects.

Also, vuepress doesnot provide any rumtime markdown parsers, you are expected to render a Vue Component and do all the stuff yourself.

**Reply by kaosine** [reply]
Well that's what we were planning on doing but didn't know how to go about doing the connects from something like axios or superclient to the router here. If I knew what I was doing, I'd just do a gha to cache the content into the site....but I was hoping there'd be an ideal way to do this.
